src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -289,45 +302,57 @@ private void EnsureResponseStream()
 
         uint statusCode;
         uint bytesSent;
-        List<GCHandle>? pinnedHeaders = SerializeHeaders(isOpaqueUpgrade);
-        try
+        using UnmanagedBufferWriter writer = new(4096);
```

Comment:
I'm not sure this can be disposed until the async operation is over.

Reply:
I thought so too, but I don't see how this worked with the previous logic then because all the GC handles were disposed of in the finally block below. I need to reexamine the logic to see what I am missing.

Reply:
Reverting my changed does make this pass, but it very odd how this works because of the aforementioned `finally` block that will always free the pinned handled and thus permit the pointers to move.

https://github.com/dotnet/aspnetcore/blob/1b53b22a866eb4bc58f2d5552db7f3ac08f1a952/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L379-L382

Another oddity is the arguments passed to `HttpSendHttpResponse`. The [official documentation](https://learn.microsoft.com/windows/win32/api/http/nf-http-httpsendhttpresponse) states the following for the `BytesSent` argument:

> Optional. A pointer to a variable that receives the number, in bytes, sent if the function operates synchronously.
>
> When making an asynchronous call using pOverlapped, set pBytesSent to NULL. Otherwise, when pOverlapped is set to NULL, pBytesSent must contain a valid memory address and not be set to NULL.

However, it is always non-NULL in the current code path, so perhaps this is a no-op? Unclear at present.

I can make this work but I'd like to hear from someone who knows this code path well. The current GCHandle approach seems to work but isn't obvious why. @Tratcher do you have any thoughts on the above?

Reply:
I started over. This was my mistake in how to deref an empty `Span<T>` and there was code that was assuming zero initialization which wasn't done automatically with the native allocator.